presto-main/src/main/java/com/facebook/presto/execution/executor/TaskExecutor.java

```diff
@@ -193,16 +210,225 @@ public TaskExecutor(TaskManagerConfig config, EmbedVersion embedVersion, Multile
+        String shuttingdownNode = gracefulShutdownSplitTracker.getNodeID();
+        log.warn("GracefulShutdown:: Going to shutdown %s tasks", currentTasksSnapshot.size());
+        for (TaskHandle taskHandle : currentTasksSnapshot) {
+            taskShutdownExecutor.execute(
+                    () -> {
+                        TaskId taskId = taskHandle.getTaskId();
+                        if (!taskHandle.getOutputBuffer().isPresent()) {
+                            log.info("No output buffer for task %s", taskId);
+                            taskHandle.handleShutDown();
+                            return;
+                        }
+
+                        OutputBuffer outputBuffer = taskHandle.getOutputBuffer().get();
+                        try {
+                            long logFrequencyMillis = 30_000;
+                            long lastLogTime = System.currentTimeMillis();  // to track when we last logged
+                            long startTime = System.nanoTime();
+                            taskHandle.updateTaskShutdownState(TaskShutdownStats.builder(shuttingdownNode).build());
+                            log.info("Output buffer for task %s= %s", taskId, taskHandle.getOutputBuffer().get().getInfo());
+                            taskHandle.updateTaskShutdownState(
+                                    builderWithOutputBufferInfo("init", shuttingdownNode, outputBuffer)
+                                            .build());
+
+                            while (taskHandle.getRunningLeafSplits() > 0 || taskHandle.getRunningIntermediateSplits() > 0) {
+                                try {
+                                    TaskShutdownStats waitingForSplitStats = builderWithOutputBufferInfo(SPLIT_WAIT, shuttingdownNode, outputBuffer)
+                                            .setPendingRunningSplitState(SPLIT_WAIT, System.nanoTime() - startTime)
+                                            .build();
+                                    taskHandle.updateTaskShutdownState(waitingForSplitStats);
+                                    long currentTime = System.currentTimeMillis();
+                                    if (currentTime - lastLogTime >= logFrequencyMillis) {
+                                        log.info("Num running splits for task %s = %s, Num blocked splits = %s", taskId, runningSplits.size(), blockedSplits.size());
+                                        logRunningWaitingAndBlockedSplits(String.format("SplitView:state:%s for task %s", SPLIT_WAIT, taskId), taskId);
+                                    }
+                                    Thread.sleep(waitTimeMillis);
+                                }
+                                catch (Exception ex) {
+                                    log.error(ex, "GracefulShutdown got interrupted while waiting for split completion for task %s", taskId);
+                                }
```

Comment:
If the intend is to catch Interrupted Exception, we need to specifically catch that and set the interrupt flag back.